src/features/hostapp/hooks/target-hostapp.ts

```diff
@@ -202,58 +211,110 @@ async function getOSReleaseResource(
+						semver_prerelease: parsedOsVersion.prerelease.join('.'),
+						semver_build: parsedOsVersion.build.join('.'),
+						revision: getRevisionFromSemver(parsedOsVersion),
+						// The OS release has to either have a matching variant,
+						// or have no variant (when it's an invariant/unified release).
+						$or: [{ variant: osVariant }, { variant: '' }],
 					},
 					{
-						$or: [
+						// We still need to be filtering hostApp releases based on the deprecated release_tags,
+						// since the versioning format of balenaOS [2019.10.0.dev, 2022.01.0] was non-semver compliant
+						// and they were not migrated to the release semver fields.
+						$and: [
 							{
 								release_tag: {
 									$any: {
 										$alias: 'rt',
 										$expr: {
 											rt: {
-												value: normalizeVariant(osVariant),
-												tag_key: 'variant',
+												// We can't use the result from balena-semver for this filter,
+												// b/c balena-semver normalizes invalid semvers like
+												// 2022.01.0 to 2022.1.0 and that would no longer
+												// match the tag value.
+												value: normalizeOsVersion(osVersion),
+												tag_key: 'version',
 											},
 										},
 									},
 								},
 							},
 							{
-								$not: {
-									release_tag: {
-										$any: {
-											$alias: 'rt',
-											$expr: {
-												rt: {
-													tag_key: 'variant',
+								// The OS release has to either have a matching variant,
+								// or have no variant release_tag (when it's an invariant/unified release).
+								$or: [
+									{
+										release_tag: {
+											$any: {
+												$alias: 'rt',
+												$expr: {
+													rt: {
+														value: normalizeVariantToLongForm(osVariant),
+														tag_key: 'variant',
+													},
 												},
 											},
 										},
 									},
-								},
+									{
+										$not: {
+											release_tag: {
+												$any: {
+													$alias: 'rt',
+													$expr: {
+														rt: {
+															tag_key: 'variant',
+														},
+													},
+												},
+											},
+										},
+									},
+								],
 							},
 						],
 					},
 				],
 			},
+			$orderby: {
+				// We order the resuts by semver_major DESC so that we always prefer rows
+				// that were matched via the semver fields rather than tags in case of a conflict,
+				// (since versions using only tags would have a 0.0.0 semver).
+				semver_major: 'desc',
+			},
 		},
-	});
+	})) as [PickDeferred<Release, 'id'>?, PickDeferred<Release, 'id'>?];
+	if (releases.length > 1) {
+		ThisShouldNeverHappenError(
+			`Found more than one hostApp release matching version ${osVersion} ${osVariant} for device type ${deviceTypeId}.`,
+		);
+	}
+
+	const [release] = releases;
+	return release;
 }
 
-function normalizeVariant(variant: string) {
+function getRevisionFromSemver(parsedOsVersion: SemVer) {
```

Comment:
Personal opinion: I strongly prefer typing function returns instead of letting the internals implicitly type the return value. This helps to reduce accidentally breaking things, and also makes the typing more explicit and intentional.

Reply:
The TS team advises to let inference do its work in most cases, but I do see your point. I've typed this and moved the defaulting to `0` outside of the function since on second thought, it isn't this helper's responsibility.